### PR TITLE
feat: Gumbel softmax and output activation interface

### LINKED ISSUE #123: [FEAT] Create an interface layer connecting categorical distributions to specific Gumbel-Softmax layers
### Interface layer implementation
To process the logits outputs of a Generator correctly, information about the type of features being synthesized is necessary.
The idea of this layer is to leverage information available in the DataProcessor of the synthesizer in order to create Gumbel-Softmax activations for each of the categorical distributions, TanH/ReLU activations for other features and concatenating all outputs in the original order.
In the end this layer should be initialized leveraging just attributes of the processor and work as a model component like any other TF/Keras layer.

Requirements:
- TF implementation
- Gumbel-Softmax layer implementation (see below)
- Tests

### Gumbel-Softmax layer implementation
A Gumbel-Softmax layer works on logits or probability outputs of a model over the distribution of a categorical feature.
The required computations of this layer's forward method are the following:
- Soft sample: Softmax of a tensor consisting of a Gumbel sample added to logits. The gradients of this computation are stored.
- Hard sample: A one-hot encoded categorical variable with a stochastically sampled class activated. The gradients of this computation are ***not*** stored, but used only for obtaining _real_ categorical samples.

Note: Generally only the hard sample should be returned in the forward call, alternatively both can be returned in an (hard, soft) tuple.

Requirements:
- TF implementation
- A standalone custom layer implemented in utils that produces the requested output assuming as input logits from a single categorical distribution
- Tests

### PR DESCRIPTION
**Pull request by jfsantos-ds**
Closes #123 
Adds a plain Gumbel-Softmax layer and a custom layer that interfaces between each output feature of the generator and the applicable activation function (Gumbel-Softmax or TanH).
This layer is initialized by passing the synthesizer's data processor object which is leveraged in order to index the different column types.
The categorical part of the output of the layer is a "sensible" one hot vector instead of a tensor of floats.
The output of the layer can also be applied directly on the inverse_transform method of the data processor.

### CODE REVIEW

File: src/ydata_synthetic/utils/gumbel_softmax.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+class ActivationInterface(Layer):
+    """An interface layer connecting different parts of an incoming tensor to adequate activation functions.
+    The tensor parts are qualified according to the passed processor object.
+    Processed categorical features are sent to specific Gumbel-Softmax layers.
+    Processed features of different kind are sent to a TanH activation.
+    Finally all output parts are concatenated and returned in the same order.
+
+    The parts of an incoming tensor are qualified by leveraging a data processor object (from the synthesizer)."""
+
+    def __init__(self, processor: BaseProcessor, name: Optional[str] = None):
```
Comment (fabclmnt) [inline src/ydata_synthetic/utils/gumbel_softmax.py:48]: Why receiving the processor? 

Please assume or create a metadata kind of object where this kind of information can be stored.

Comment (jfsantos-ds) [reply src/ydata_synthetic/utils/gumbel_softmax.py]: Added a metadata object to the processor to be leveraged here and in this kind of developments!
